### PR TITLE
Mapping performance improvements

### PR DESCRIPTION
**Pull request by grzesiek2010**
Work towards #5087 

#### What has been done to verify that this works as intended?
I've tested the performance manually.

#### Why is this the best possible solution? Were any other approaches considered?
This pr contains three improvements:
1. Caching bitmaps (markers) in mapbox - this is what we already had for google so it seems like a nice improvement in mapbox too to avoid huge memory usage.
2. Fix for setting marker icons twice - before we would create every marker with a default icon and then set what made the process longer.
3. Fixed highlighting already selected items after rotating a device.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Displaying icons for items on a map might be affected so this is what should be tested (map of forms and choices). Nothing else should be affected.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: maps/src/main/java/org/odk/collect/maps/MapFragment.java [hunk 1]
```diff
@@ -123,7 +123,7 @@ void addTo(
      * By default, the marker icon will be R.drawable.ic_map_point.
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/MapFragment.java]: This comment needs updated (or have an overloaded version that uses the "default" described in the comment).

File: geo/src/main/java/org/odk/collect/geo/selection/SelectionMapFragment.kt [hunk 2]
```diff
@@ -354,6 +350,9 @@ class SelectionMapFragment(
             itemsByFeatureId[featureId] = item
             points.add(point)
         }
+        selectedFeatureViewModel.getSelectedFeatureId()?.let {
```
Comment (seadowg) [inline geo/src/main/java/org/odk/collect/geo/selection/SelectionMapFragment.kt]: It looks like we're missing test coverage for this change (and the original placement of the code). I guess the code itself is for two things:

1. Keeping the currently selected item selected if the items change (not really a use case we currently have)
2. Keeping the currently selected item selected when the Fragment is recreated

I've added a test for **2** and verified that it forces us to move the code around by moving it back and seeing it fail.

In writing that test, I then realized there was a problem where if we started the fragment with a selection, selecting a new item and then selecting again would always result in the initial selection being displayed. I wrote a test for that case and then realized we don't need this code at all. That's all committed here.

**Review by seadowg — APPROVED** [review]

File: maps/src/main/java/org/odk/collect/maps/MapFragment.java (hunk 1 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/MapFragment.java]: Fixed.

**Comment by lognaturel** [comment]
@seadowg I am overjoyed to see you add PRs to milestones. 🎉😅

**Comment by seadowg** [comment]
> I am overjoyed to see you add PRs to milestones. 🎉😅

Ah because it doesn't have an attached issue!

**Comment by lognaturel** [comment]
> Ah because it doesn't have an attached issue!

I see, makes sense. At some point in time I remember @grzesiek2010 was removing an issue from the milestone and putting the PR in so there wasn't redundancy and the most "active thing" was visible in the milestone. That worked well for me though it is annoying to do by hand. Ah well, by the time we figure it out I'm sure GitHub will have a concept that makes both the issue and PR get visually liked to the milestone.

**Comment by srujner** [comment]
**Tested with Success!**
Verified on Android 10 and 11

Verifies cases:
- View submissions on map;
- Google Maps, Mapbox, OSM, use offline layer;
- Regression checks on geowidgets, background location;
- Performance on Thousand points geojson widgets using Mapbox;
- Highlighting already selected items after rotating a device;
- Displaying icons for items on a map;
- Light and Dark mode;
- Screen rotation, minimize app, using backward button.

**Comment by kkrawczyk123** [comment]
Tested also on Androids: 5.1 and 8.1

I have also noticed another small screen rotation issue and reported in #5113

**Pull request merged**